Add a release checklist to guide and standardize the release process (#429)
Yoast/yoastcs: .github/RELEASE_CHECKLIST.md

Release checklist
===========================================================

## Pre-release

- [ ] If no upcoming versioned milestone exists, and there are merged PRs in the `Next release` milestone:
    - [ ] Determine what the version number should be for the release and rename the `Next release` milestone to that version number.
      This project uses [semantic versioning](https://semver.org/) to determine the version of the next release.

>     - [ ] Move anything that is still open in the `Next release` milestone to a new `Next release` milestone.

Reviewer: ```suggestion
    - [ ] Create a new "Next release" milestone for the release _after_ this one.
    - [ ] Move anything that is still open in the current milestone to the new `Next release` milestone.
```

- [ ] Verify that any PR merged since the last release has a milestone attached to it.
- [ ] Create a changelog based on everything in the milestone, pull the changelog changes & merge the PR.
  Verify that a release link at the bottom of the `CHANGELOG.md` file has been added.
- [ ] Make sure that `develop` is ahead of `main`.

>   - If not, merge `main` into `develop` or fast-forward `develop` to be equal to `main`.

Reviewer: ```suggestion
- [ ] Double-check that nothing was merged into the `main` branch.
    - If there are unreleased commits in `main`, merge `main` into `develop` or cherrypick the commits from main into the `develop` branch.
```

Reasoning: `develop` is where things are normally merged to, so a "fast-forward `develop` to be equal to `main`" will never work.

## Release

- [ ] Open a PR to merge the `develop` branch to `main`.
- [ ] Merge that PR once all CI checks have passed.

> - [ ] Create a release on GitHub, also creating a new tab (targeting `main`) in the process, and copy & paste the changelog to it.

Reviewer: ```suggestion
- [ ] Create a release on GitHub, also creating a new tag (targeting `main`) in the process, and copy & paste the changelog to it.
```

>   - Add/edit the relevant links to the bottom of the changelog to make sure all texts between `[]` in changelog entries become links.

Reviewer: ```suggestion
    - Add/edit the relevant links to the bottom of the changelog to make sure all texts between `[]` in changelog entries become links.
```

    - For example: `[PHP_CodeSniffer]` only becomes a link when `[PHP_CodeSniffer]: https://github.com/squizlabs/PHP_CodeSniffer/releases` is present as well.
- [ ] Publish the release.

> - [ ] Merge the `main` branch into `develop`.

Reviewer: ```suggestion
- [ ] Fast-forward `develop` to `main`.
```

Reasoning: here we can do a fast-forward as we just merged the PR from `develop` into `main` a few steps earlier.

- [ ] Close the milestone.
- [ ] Open a new milestone for the next release.

> - [ ] If any open PRs/issues which were milestoned for the release did not make it into the release, update their milestone.

Reviewer: ```suggestion
```

These steps are already covered in "Pre-release".

Release checklist
===========================================================

## Pre-release

> - [ ] If no upcoming versioned milestone exists:

Reviewer: ```suggestion
- [ ] If no upcoming versioned milestone exists and there are merged PRs in the `Next release` milestone:
```

    - [ ] Determine what the version number should be for the release and rename the `Next release` milestone to that version number.
        This project uses [semantic versioning](https://semver.org/) to determine the version of the next release.
    - [ ] Move anything that is still open in the `Next release` milestone to a new `Next release` milestone.

> 

Reviewer: ```suggestion
- [ ] Verify that any PR merged since the last release has a milestone attached to it.
- [ ] Create a changelog based on everything in the milestone, pull the changelog changes & merge the PR.

```

## Release

- [ ] Checkout `develop`:
    ```bash
    git checkout develop
    ```
- [ ] Make sure the branch is up to date with `main`:
    ```bash
    git merge origin/main

>     ```

Reviewer: ```suggestion
- [ ] Check out the `main` branch and fast-forward the `main` branch to be at the same commit as `develop` & push:
    ```bash
    git checkout main
    git merge origin/develop
    git push origin main
    ```
```

I realize the net-effect will be the same, but it makes more sense to merge `develop` into `main` for a release, than to merge an unchanged `main` into `develop`. That's just doing things the wrong way round.

Alternatively - and I believe a variation of this workflow was followed up to the 3.0.0 release:
```suggestion
- [ ] Open a PR to merge the `develop` branch to `main`.
- [ ] Merge that PR once all CI checks have passed.
```

Author: The idea here was to make sure that `develop` is ahead of `main` in case for any reason a commit has landed directly into `main`. I think we can just mention that in the prerelease checks instead.

Reviewer: > The idea here was to make sure that develop is ahead of main in case for any reason a commit has landed directly into main.

Cricky... I can't even think of how that would happen, what with the default PR branch being `develop`, but fair point. Can't say I've come across that at any time in this repo though.

> I think we can just mention that in the prerelease checks instead.

Let's 👍🏻

> - [ ] Update version, date and add changelog entries in `CHANGELOG.md`.

Reviewer: Update version ?

Also: this is not the right order - the changelog should be pulled & merged before the merge from `develop` to `main` happens.
Basically, IMO, this step should be moved to "Pre-release".

    - Guidelines: [Keep a CHANGELOG](https://keepachangelog.com/).
    - Verify that a release link at the bottom of the `CHANGELOG.md` file has been added.
- [ ] Push branch to origin:
    ```bash
    git push origin develop

>     ```

Reviewer: I suggest removing this step and joining it in the first step (as per suggestion)

- [ ] Make sure all CI builds are green.
- [ ] Check out the `main` branch:
    ```bash
    git checkout main
    ```
- [ ] Merge `develop` into `main`:
    ```bash
    git merge --no-ff develop

>     ```

Reviewer: I honestly don't know what value these steps have.

> - [ ] Create a new tag:

Reviewer: Why do this as a separate step ? Why not let GH automatically create the tag when doing the release ?

    ```bash
    git tag -a [new-version]
    ```
- [ ] Push the commit and tag to the remote:
    ```bash
    git push origin main --tags
    ```
- [ ] Create a new release on GitHub and add the changelog entry to it.
    - Add the relevant links to the bottom of the changelog to make sure all texts between `[]` in changelog entries become links. 
      - For example: `[PHP_CodeSniffer]` only becomes a link when `[PHP_CodeSniffer]: https://github.com/squizlabs/PHP_CodeSniffer/releases` is present as well.
- [ ] Merge `main` into `develop`:
    ```bash
    git checkout develop
    git merge origin/main
    git push origin develop

>     ```

Reviewer: Shouldn't be needed anymore if my other suggestions are accepted.

- [ ] Close the milestone.
- [ ] Create a new `Next release` milestone if it does not yet exist.

> - [ ] If any open PRs/issues that were milestoned for the release did not make it into the release, update their milestone.

Reviewer: These steps are already in the "Pre-release" checklist and don't need to be duplicated here (though the text of what's in "Pre-release" may need to be improved/clarified).
